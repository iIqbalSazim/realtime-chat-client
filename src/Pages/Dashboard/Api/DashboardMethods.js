import { getRequest, postRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (receiverId) => {
  return await getRequest(`messages?chatting_with=${receiverId}`);
};

export const sendMessage = async (message) => {
  return await postRequest("messages", message);
};
