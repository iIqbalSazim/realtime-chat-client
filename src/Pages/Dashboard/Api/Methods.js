import { getRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (senderId, receiverId) => {
  return await getRequest(
    `messages?sender_id=${senderId}&receiver_id=${receiverId}`
  );
};
