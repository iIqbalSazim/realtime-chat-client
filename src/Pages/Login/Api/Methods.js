import { postRequest } from "../../../Config/Axios/AxiosConfig";

export const loginUser = async (user) => {
  return await postRequest("login", user);
};

export const revokeToken = async (token) => {
  return await postRequest("revoke_token", token);
};
