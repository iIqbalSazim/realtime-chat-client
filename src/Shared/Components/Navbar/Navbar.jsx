import { useDispatch, useSelector } from "react-redux";
import { Link, useNavigate } from "react-router-dom";

import { logout } from "../../../Stores/Actions/auth";
import { clearChatUser, setMessages } from "../../../Stores/Actions/chat";
import { revokeToken } from "../../../Pages/Login/Api/LoginMethods";

const Navbar = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const currentUser = useSelector((state) => state.auth.user);

  const handleLogout = async () => {
    const token = localStorage.getItem("token");

    dispatch(clearChatUser());
    dispatch(logout());
    dispatch(setMessages([]));
    navigate("/");
    await revokeToken({ token });
    localStorage.removeItem("token");
  };

  return (
    <nav className="flex justify-center gap-5 p-2 absolute z-50 right-2">
      <Link
        to={`/${currentUser.id}`}
        className="btn btn-sm btn-secondary btn-outline"
      >
        Dashboard
      </Link>
      <button onClick={() => handleLogout()} className="btn btn-sm btn-accent">
        Logout
      </button>
    </nav>
  );
};

export default Navbar;
