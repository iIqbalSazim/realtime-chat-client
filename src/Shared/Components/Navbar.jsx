import { useDispatch, useSelector } from "react-redux";
import { Link, useNavigate } from "react-router-dom";

import { logout } from "../../Stores/Actions/auth";
import { clearChatUser, setMessages } from "../../Stores/Actions/chat";
import { revokeToken } from "../../Pages/Login/Api/Methods";

const Navbar = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const currentUser = useSelector((state) => state.auth.user);

  const handleLogout = async () => {
    dispatch(clearChatUser());
    dispatch(logout());
    dispatch(setMessages([]));
    const token = localStorage.getItem("token");
    navigate("/");
    localStorage.removeItem("token");
    revokeToken({ token });
  };

  return (
    <nav className="flex justify-center gap-5 p-2 absolute z-50 right-2">
      <Link
        to={`/${currentUser.id}`}
        className="btn btn-sm btn-secondary btn-outline"
      >
        Dashboard
      </Link>
      <button onClick={() => handleLogout()} className="btn btn-sm btn-accent">
        Logout
      </button>
    </nav>
  );
};

export default Navbar;
