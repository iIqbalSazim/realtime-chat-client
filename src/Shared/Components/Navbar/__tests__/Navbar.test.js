import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../../../../jest.setup";

import * as ApiMethods from "../../../../Pages/Login/Api/LoginMethods";

import Navbar from "../Navbar";

describe("Navbar", () => {
  test("renders correctly", () => {
    jest.spyOn(require("react-redux"), "useSelector").mockReturnValue({
      id: 1,
    });
    const { container } = renderWithProviders(<Navbar />);

    const dashboardLink = screen.getByRole("link", { name: "Dashboard" });
    const logoutButton = screen.getByRole("button", { name: "Logout" });

    expect(dashboardLink).toBeInTheDocument();
    expect(logoutButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("calls revokeToken function upon clicking on the logout button", async () => {
    const revokeTokenSpy = jest.spyOn(ApiMethods, "revokeToken");

    renderWithProviders(<Navbar />);

    const logoutButton = screen.getByRole("button", { name: "Logout" });

    await userEvent.click(logoutButton);

    await waitFor(() => {
      expect(revokeTokenSpy).toBeCalled();
    });
  });
});
