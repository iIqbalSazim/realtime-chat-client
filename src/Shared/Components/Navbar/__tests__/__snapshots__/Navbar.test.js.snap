// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div>
  <nav
    class="flex justify-center gap-5 p-2 absolute z-50 right-2"
  >
    <a
      class="btn btn-sm btn-secondary btn-outline"
      href="/1"
    >
      Dashboard
    </a>
    <button
      class="btn btn-sm btn-accent"
    >
      Logout
    </button>
  </nav>
</div>
`;
